src/uploader.c

```diff
@@ -1,2409 +1,127 @@
-                       "memory error: unable to malloc");
-        goto clean_variables;
-    }
-
-    for (int i = 0; i < state->total_parity_shards; i++) {
-        fec_blocks[i] = map_parity + i * state->shard_size;
-    }
-
-    state->log->debug(state->env->log_options, state->handle,
-                      "Encoding parity shards, data_shards: %i, "       \
-                      "parity_shards: %i, shard_size: %" PRIu64 ", "    \
-                      "file_size: %" PRIu64,
-                      state->total_data_shards,
-                      state->total_parity_shards,
-                      state->shard_size,
-                      state->file_size);
-
-
-    reed_solomon *rs = reed_solomon_new(state->total_data_shards,
-                                        state->total_parity_shards);
-    reed_solomon_encode2(rs, data_blocks, fec_blocks, state->total_shards,
-                         state->shard_size, state->file_size);
-    reed_solomon_release(rs);
-
-clean_variables:
-    if (data_blocks) {
-        free(data_blocks);
-    }
-
-    if (fec_blocks) {
-        free(fec_blocks);
-    }
-
-    if (tmp_folder) {
-        free(tmp_folder);
-    }
-
-    if (map) {
-        unmap_file(map, state->file_size);
-    }
-
-    if (map_parity) {
-        unmap_file(map_parity, parity_size);
-    }
-
-    if (parity_file) {
-        fclose(parity_file);
-    }
-
-    if (encrypted_file) {
-        fclose(encrypted_file);
-    }
-}
-
-
-static void queue_create_parity_shards(storj_upload_state_t *state)
+static uv_work_t *uv_work_new()
 {
-    uv_work_t *work = uv_work_new();
-    if (!work) {
-        state->error_status = STORJ_MEMORY_ERROR;
-        return;
-    }
-
-    state->pending_work_count += 1;
-
-    parity_shard_req_t *req = malloc(sizeof(parity_shard_req_t));
-
-    req->error_status = 0;
-    req->upload_state = state;
-    work->data = req;
-
-    int status = uv_queue_work(state->env->loop, (uv_work_t*) work,
-                               create_parity_shards, after_create_parity_shards);
-
-    if (status) {
-        state->error_status = STORJ_QUEUE_ERROR;
-    }
-
-    state->awaiting_parity_shards = false;
+    uv_work_t *work = malloc(sizeof(uv_work_t));
+    return work;
 }
 
-static void after_send_exchange_report(uv_work_t *work, int status)
+static void cleanup_state(storj_upload_state_t *state)
 {
-    shard_send_report_t *req = work->data;
-
-    req->state->pending_work_count -= 1;
-
-    if (status == UV_ECANCELED) {
-        req->report->send_count = 0;
-        req->report->send_status = STORJ_REPORT_AWAITING_SEND;
-
-        goto clean_variables;
-    }
-
-    req->report->send_count += 1;
-
-    if (req->status_code == 201) {
-        req->state->env->log->info(req->state->env->log_options,
-                        req->state->handle,
-                         "Successfully sent exchange report for shard %d",
-                         req->report->pointer_index);
-
-        req->report->send_status
```

Comment:
`cleanup_work` in `downloader.c` does a couple extra things. We should do them here if necessary. We should also change the name of this function to `cleanup_work`, or change the name of the function in the downloader to `cleanup_download_work` to make them consistent